issue when applying changes to a `okta_user_schema`

<!---
Please note the following potential times when an issue might be in Terraform core:

* [Configuration Language](https://www.terraform.io/docs/configuration/index.html) or resource ordering issues
* [State](https://www.terraform.io/docs/state/index.html) and [State Backend](https://www.terraform.io/docs/backends/index.html) issues
* [Provisioner](https://www.terraform.io/docs/provisioners/index.html) issues
* [Registry](https://registry.terraform.io/) issues
* Spans resources across multiple providers

If you are running into one of these scenarios, we recommend opening an issue in the [Terraform core repository](https://github.com/hashicorp/terraform/) instead.
--->

<!--- Please keep this note for the community --->

### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform Version
terraform 0.14.2 
But this also happens on .13

<!--- Please run `terraform -v` to show the Terraform core version and provider version(s). If you are not running the latest version of Terraform or the provider, please upgrade because your issue may have already been fixed. [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions). --->

### Affected Resource(s)

<!--- Please list the affected resources and data sources. --->

okta_user_schema

### Terraform Configuration Files

<!--- Information about code formatting: https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code --->

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
resource "okta_user_schema" "ca_regsrc" {
  index       = "registration_source"
  title       = "Registration Source"
  type        = "string"
  description = "Name of the business unit that registered the user."
  scope       = "NONE"
  # min_length  = 1
  # max_length  = 200
  permissions = "READ_ONLY"
  master      = "PROFILE_MASTER"


}

### Debug Output

<!---
Please provide a link to a GitHub Gist containing the complete debug output. Please do NOT paste the debug output in the issue; just paste a link to the Gist.

To obtain the debug output, see the [Terraform documentation on debugging](https://www.terraform.io/docs/internals/debugging.html).
--->

### Panic Output

<!--- If Terraform produced a panic, please provide a link to a GitHub Gist containing the output of the `crash.log`. --->

### Expected Behavior

<!--- What should have happened? --->
A new custom attribute in the Okta User profile should have been created.  
It sometimes is created, sometimes this shows when running an apply against an already created custom attribute

When applying changes to okta_user_schema.ca_regsrc, provider
2020-12-12T01:22:20.4152573Z "registry.terraform.io/terraform-providers/okta" produced an unexpected new
2020-12-12T01:22:20.4154031Z value: Root resource was present, but now absent.

<!--- What actually happened? --->

### Steps to Reproduce

Run terraform apply to an org and the error happens for me

1. `terraform apply`

### Important Factoids

<!--- Are there anything atypical about your accounts that we should know? For example: Running in EC2 Classic? --->

### References

<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Vendor documentation? For example:
--->

- #0000

I tried to revert to version 13.1, but the error is still there.  This time it seemed to create one of the custom attributes but not the other.  One custom attribute was created with no error.
2020-12-14T17:15:36.8118933Z [1m[31mError: [0m[0m[1mProvider produced inconsistent result after apply[0m
2020-12-14T17:15:36.8119848Z 
2020-12-14T17:15:36.8120431Z [0mWhen applying changes to okta_user_schema.ca_regsrc, provider
2020-12-14T17:15:36.8121848Z "registry.terraform.io/terraform-providers/okta" produced an unexpected new
2020-12-14T17:15:36.8122443Z value: Root resource was present, but now absent.
2020-12-14T17:15:36.8122733Z 
2020-12-14T17:15:36.8124996Z This is a bug in the provider, which should be reported in the provider's own
2020-12-14T17:15:36.8125803Z issue tracker.
2020-12-14T17:15:36.8126059Z [0m[0m
2020-12-14T17:15:36.8126293Z [31m
2020-12-14T17:15:36.8126895Z [1m[31mError: [0m[0m[1mProvider produced inconsistent result after apply[0m
2020-12-14T17:15:36.8127328Z 
2020-12-14T17:15:36.8129394Z [0mWhen applying changes to okta_user_schema.ca_cci_master, provider
2020-12-14T17:15:36.8131862Z "registry.terraform.io/terraform-providers/okta" produced an unexpected new
2020-12-14T17:15:36.8134053Z value: Root resource was present, but now absent.
2020-12-14T17:15:36.8134409Z 
2020-12-14T17:15:36.8136504Z This is a bug in the provider, which should be reported in the provider's own
2020-12-14T17:15:36.8139230Z issue tracker.
2020-12-14T17:15:36.8139366Z [0m[0m
2020-12-14T17:15:37.4464002Z ##[error]PowerShell exited with code '1'.

--- a/okta/resource_okta_policy_password.go
+++ b/okta/resource_okta_policy_password.go
@@ -215,7 +215,9 @@ func resourcePolicyPasswordRead(ctx context.Context, d *schema.ResourceData, m i
 		_ = d.Set("password_min_number", policy.Settings.Password.Complexity.MinNumber)
 		_ = d.Set("password_min_symbol", policy.Settings.Password.Complexity.MinSymbol)
 		_ = d.Set("password_exclude_username", policy.Settings.Password.Complexity.ExcludeUsername)
-		_ = d.Set("password_dictionary_lookup", policy.Settings.Password.Complexity.Dictionary.Common.Exclude)
+		if policy.Settings.Password.Complexity.Dictionary != nil && policy.Settings.Password.Complexity.Dictionary.Common != nil {
+			_ = d.Set("password_dictionary_lookup", policy.Settings.Password.Complexity.Dictionary.Common.Exclude)
+		}
 		_ = d.Set("password_max_age_days", policy.Settings.Password.Age.MaxAgeDays)
 		_ = d.Set("password_expire_warn_days", policy.Settings.Password.Age.ExpireWarnDays)
 		_ = d.Set("password_min_age_minutes", policy.Settings.Password.Age.MinAgeMinutes)

--- a/okta/resource_okta_user_schema.go
+++ b/okta/resource_okta_user_schema.go
@@ -2,8 +2,9 @@ package okta
 
 import (
 	"context"
-	"fmt"
+	"reflect"
 	"strings"
+	"time"
 
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
@@ -14,9 +15,9 @@ const customSchema = "custom"
 
 func resourceUserSchema() *schema.Resource {
 	return &schema.Resource{
-		CreateContext: resourceUserSchemaCreate,
+		CreateContext: resourceUserSchemaCreateOrUpdate,
 		ReadContext:   resourceUserSchemaRead,
-		UpdateContext: resourceUserSchemaUpdate,
+		UpdateContext: resourceUserSchemaCreateOrUpdate,
 		DeleteContext: resourceUserSchemaDelete,
 		Importer: &schema.ResourceImporter{
 			StateContext: func(ctx context.Context, d *schema.ResourceData, m interface{}) ([]*schema.ResourceData, error) {
@@ -57,28 +58,59 @@ func resourceUserSchemaResourceV0() *schema.Resource {
 	return &schema.Resource{Schema: buildSchema(userBaseSchemaSchema, userSchemaSchema)}
 }
 
-func resourceUserSchemaCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+// Sometime Okta API does not update or create custom property on the first try, thus that require running
+// `terraform apply` several times. This simple retry resolves that issue. (If) After  this issue will be resolved,
+// this retry logic will be demolished.
+func resourceUserSchemaCreateOrUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	logger(m).Info("creating user schema", "name", d.Get("index").(string))
 	schemaUrl, err := getUserTypeSchemaUrl(ctx, getOktaClientFromMetadata(m), d.Get("user_type").(string))
 	if err != nil {
 		return diag.Errorf("failed to create user custom schema: %v", err)
 	}
-	_, _, err = getSupplementFromMetadata(m).UpdateCustomUserSchemaProperty(ctx, schemaUrl, d.Get("index").(string), userSubSchema(d))
-	if err != nil {
-		return diag.Errorf("failed to create user custom schema: %v", err)
+	var subschema *sdk.UserSubSchema
+	timer := time.NewTimer(time.Second * 3)
+	ticker := time.NewTicker(time.Millisecond * 500)
+loop:
+	for {
+		select {
+		case <-ctx.Done():
+			return diag.Errorf("failed to create user custom schema: %v", ctx.Err())
+		case <-timer.C:
+			return diag.Errorf("failed to create user custom schema: no more attempts left")
+		case <-ticker.C:
+			updated, _, err := getSupplementFromMetadata(m).UpdateCustomUserSchemaProperty(ctx, schemaUrl, d.Get("index").(string), userSubSchema(d))
+			if err != nil {
+				return diag.Errorf("failed to create user custom schema: %v", err)
+			}
+			d.SetId(d.Get("index").(string))
+			s, _, err := getSupplementFromMetadata(m).GetUserSchema(ctx, schemaUrl)
+			if err != nil {
+				return diag.Errorf("failed to get user custom schema: %v", err)
+			}
+			subschema = getCustomProperty(s, d.Id())
+			if subschema != nil && reflect.DeepEqual(subschema, updated.Definitions.Custom.Properties[d.Id()]) {
+				break loop
+			}
+		}
 	}
-	d.SetId(d.Get("index").(string))
-	return resourceUserSchemaRead(ctx, d, m)
+	err = syncUserSchema(d, subschema)
+	if err != nil {
+		return diag.Errorf("failed to set user custom schema properties: %v", err)
+	}
+	return nil
 }
 
 func resourceUserSchemaRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	logger(m).Info("reading user schema", "name", d.Get("index").(string))
 	schemaUrl, err := getUserTypeSchemaUrl(ctx, getOktaClientFromMetadata(m), d.Get("user_type").(string))
 	if err != nil {
 		return diag.Errorf("failed to get user custom schema: %v", err)
 	}
-	subschema, err := getSubSchema(ctx, getSupplementFromMetadata(m), schemaUrl, d)
+	s, _, err := getSupplementFromMetadata(m).GetUserSchema(ctx, schemaUrl)
 	if err != nil {
 		return diag.Errorf("failed to get user custom schema: %v", err)
 	}
+	subschema := getCustomProperty(s, d.Id())
 	if subschema == nil {
 		d.SetId("")
 		return nil
@@ -88,26 +120,6 @@ func resourceUserSchemaRead(ctx context.Context, d *schema.ResourceData, m inter
 		return diag.Errorf("failed to set user custom schema properties: %v", err)
 	}
 	return nil
-}
-
-func getSubSchema(ctx context.Context, client *sdk.ApiSupplement, schemaUrl string, d *schema.ResourceData) (*sdk.UserSubSchema, error) {
-	s, _, err := client.GetUserSchema(ctx, schemaUrl)
-	if err != nil {
-		return nil, fmt.Errorf("failed to get user custom schema: %v", err)
-	}
-	return getCustomProperty(s, d.Id()), err
-}
-
-func resourceUserSchemaUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
-	schemaUrl, err := getUserTypeSchemaUrl(ctx, getOktaClientFromMetadata(m), d.Get("user_type").(string))
-	if err != nil {
-		return diag.Errorf("failed to update user custom schema: %v", err)
-	}
-	_, _, err = getSupplementFromMetadata(m).UpdateCustomUserSchemaProperty(ctx, schemaUrl, d.Get("index").(string), userSubSchema(d))
-	if err != nil {
-		return diag.Errorf("failed to update user custom schema: %v", err)
-	}
-	return resourceUserSchemaRead(ctx, d, m)
 }
 
 func resourceUserSchemaDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {

--- a/okta/resource_okta_user_schema_test.go
+++ b/okta/resource_okta_user_schema_test.go
@@ -255,19 +255,12 @@ func testSchemaPropertyExists(schemaUserType, index, resolutionScope string) (bo
 	}
 	switch resolutionScope {
 	case baseSchema:
-		for key := range s.Definitions.Base.Properties {
-			if key == index {
-				return true, nil
-			}
-		}
+		_, ok := s.Definitions.Base.Properties[index]
+		return ok, nil
 	case customSchema:
-		for key := range s.Definitions.Custom.Properties {
-			if key == index {
-				return true, nil
-			}
-		}
+		_, ok := s.Definitions.Custom.Properties[index]
+		return ok, nil
 	default:
-		return false, fmt.Errorf("resolution scope can be 'base' or 'custom'")
+		return false, fmt.Errorf("resolution scope can be only 'base' or 'custom'")
 	}
-	return false, nil
 }

--- a/okta/user_schema.go
+++ b/okta/user_schema.go
@@ -210,28 +210,18 @@ func syncBaseUserSchema(d *schema.ResourceData, subschema *sdk.UserSubSchema) {
 	}
 }
 
-func getBaseProperty(us *sdk.UserSchema, id string) *sdk.UserSubSchema {
-	if us == nil {
+func getBaseProperty(s *sdk.UserSchema, id string) *sdk.UserSubSchema {
+	if s == nil {
 		return nil
 	}
-	for key, part := range us.Definitions.Base.Properties {
-		if key == id {
-			return part
-		}
-	}
-	return nil
+	return s.Definitions.Base.Properties[id]
 }
 
 func getCustomProperty(s *sdk.UserSchema, id string) *sdk.UserSubSchema {
 	if s == nil {
 		return nil
 	}
-	for key, part := range s.Definitions.Custom.Properties {
-		if key == id {
-			return part
-		}
-	}
-	return nil
+	return s.Definitions.Custom.Properties[id]
 }
 
 func getNullableOneOf(d *schema.ResourceData, key string) (oneOf []*sdk.UserSchemaEnum) {

